src/backend/cdb/cdbpath.c

```diff
@@ -224,24 +224,36 @@ cdbpath_create_motion_path(PlannerInfo *root,
 			return subpath;
 		}
 
-		if (CdbPathLocus_IsGeneral(subpath->locus))
-		{
-			/* XXX: this is a bit bogus. We just change the subpath's locus. */
-			subpath->locus = locus;
-			return subpath;
-		}
-
-		if (CdbPathLocus_IsEntry(subpath->locus) ||
+		if (CdbPathLocus_IsGeneral(subpath->locus) ||
+			CdbPathLocus_IsEntry(subpath->locus) ||
```

Comment:
1. May the path of General locus contain motions?
2. Is it valid to alter Entry/SingleQE data flow to Outer always? They will be able to execute on multiple segments as Outer locus allows.

Reply:
My patch does this for cases, when there is just one Motion node or Projection path and Motion node (and it also adds Materialize node, as Outer locus requires). For other cases new Materialize and Motion nodes are added at the end of function.